src/dotty/tools/dotc/typer/Applications.scala

```diff
@@ -1146,11 +1146,17 @@ trait Applications extends Compatibility { self: Typer =>
             alts
         }
 
-        def narrowByTrees(alts: List[TermRef], args: List[Tree], resultType: Type): List[TermRef] =
-          alts filter ( alt =>
-            if (!ctx.isAfterTyper) isApplicable(alt, targs, args, resultType)
-            else isDirectlyApplicable(alt, targs, args, resultType)
+        def narrowByTrees(alts: List[TermRef], args: List[Tree], resultType: Type): List[TermRef] = {
+          val alts2 = alts.filter(alt =>
+            isDirectlyApplicable(alt, targs, args, resultType)
           )
+          if (alts2.isEmpty && !ctx.isAfterTyper)
+            alts.filter(alt =>
+              isApplicable(alt, targs, args, resultType)
+            )
+          else
+            alts2
```

Comment:
From the commit message of https://github.com/lampepfl/dotty/pull/1288/commits/7a060ba8ed52bc4a5762917c8f0cb1322542bc4a:
> If a directly applicable alternative exists, there is no need to try
every alternative (which would require searching for implicit
conversions for every argument that doesn't directly match). This should
not affect semantics since applicable overloads that require some
implicit conversions are dismissed in `narrowMostSpecific`

The motivation for this change is that I observed that many of the implicit searches when compiling dotty were caused by overloading resolution (e.g., every call to `Int#+` will check if the argument can be converted to `Char`,`Short`,`Double`, etc even if the argument is an `Int`)  It's possible that I missed a case where this does change the semantics, if so could you give an example?